packages/headless/src/controllers/commerce/instant-products/headless-instant-products.ts

```diff
@@ -0,0 +1,184 @@
+import {NumberValue, Schema} from '@coveo/bueno';
+import {SerializedError} from '@reduxjs/toolkit';
+import {CommerceAPIErrorResponse} from '../../../api/commerce/commerce-api-error-response';
+import {Product} from '../../../api/commerce/common/product';
+import {CommerceEngine} from '../../../app/commerce-engine/commerce-engine';
+import {
+  clearExpiredProducts,
+  registerInstantProducts,
+  updateInstantProductsQuery,
+} from '../../../features/commerce/instant-products/instant-products-actions';
+import {instantProductsReducer} from '../../../features/commerce/instant-products/instant-products-slice';
+import {fetchInstantProducts} from '../../../features/commerce/search/search-actions';
+import {hasExpired} from '../../../features/instant-items/instant-items-state';
+import {InstantProductsSection} from '../../../state/state-sections';
+import {loadReducerError} from '../../../utils/errors';
+import {randomID} from '../../../utils/utils';
+import {nonEmptyString, validateOptions} from '../../../utils/validate-payload';
+import {
+  Controller,
+  buildController,
+} from '../../controller/headless-controller';
+
+export interface InstantProductsOptions {
+  /**
+   * A unique identifier for the search box.
+   */
+  searchBoxId?: string;
+  /**
+   * Number in milliseconds that cached products will be valid for. Defaults to 1 minute. Set to 0 so that products never expire.
+   */
+  cacheTimeout?: number;
+}
+
+const instantProductsOptionDefinitions = {
+  searchBoxId: nonEmptyString,
+  cacheTimeout: new NumberValue(),
+};
+
+const instantProductsOptionsSchema = new Schema<
+  Required<InstantProductsOptions>
+>(instantProductsOptionDefinitions);
+
+export interface InstantProductsProps {
+  options: InstantProductsOptions;
+}
+
+/**
+ * The `InstantProducts` controller allows the end user to manage instant products queries.
+ */
+export interface InstantProducts extends Controller {
+  /**
+   * Updates the specified query and shows instant products for it.
+   *
+   * @param q - The query to get instant products for. For more precise instant products, query suggestions are recommended.
+   */
+  updateQuery(q: string): void;
+  /**
+   * Clears all expired instant products queries.
+   */
+  clearExpired(): void;
+  /**
+   * The state of the `InstantProducts` controller.
+   */
+  state: InstantProductsState;
+}
+
+export interface InstantProductsState {
+  /**
+   * The current query for instant products.
+   */
+  q: string;
```

Comment:
We've used `query` across commerce controllers, since it's also the name of the parameter on our API. If it makes it easier to re-use existing actions, I don't mind leaving it as-is.
```suggestion
  query: string;
```